File: test/ts-stub-result-overlay/src/wiring.test.ts
import {BaboonBinWriter, BaboonMethodId} from "./baboon_runtime";
import {IBaboonServiceRt} from "./generated/testpkg/pkg0/baboon-service-rt";
import {Result, ResultServiceRt} from "./custom-containers";
import {In} from "./generated/testpkg/pkg0/i1/testcall/in";
import {In as In_No_Err} from "./generated/testpkg/pkg0/i2/noerrcall/in";
import {Out} from "./generated/testpkg/pkg0/i1/testcall/out";
import {Out as Out_No_Err} from "./generated/testpkg/pkg0/i2/noerrcall/out";
import {Err} from "./generated/testpkg/pkg0/i1/testcall/err";
import {I2} from "./generated/testpkg/pkg0/I2";
import {I1} from "./generated/testpkg/pkg0/I1";
import {BaboonBinReader, BaboonCodecContext} from "./generated/BaboonSharedRuntime";
import {invokeJson_I1, invokeUeba_I1} from "./generated/testpkg/pkg0/i1-wiring";
import {T7_Empty} from "./generated/testpkg/pkg0/T7_Empty";
import {invokeJson_I2, invokeUeba_I2} from "./generated/testpkg/pkg0/i2-wiring";

const ctx = BaboonCodecContext.Default;
const rt: IBaboonServiceRt = ResultServiceRt;

const mockI1: I1 = {
    testCall(arg: In): Result<Out, Err> {
        return {tag: "Success", value: new Out(42)};
    },
    testCall2(arg: T7_Empty): Result<T7_Empty, T7_Empty> {
        return {tag: "Success", value: new T7_Empty()};
    },
};

const failingI1: I1 = {
    testCall(arg: In): Result<Out, Err> {
        return {tag: "Failure", error: new Err("domain error")};
    },
    testCall2(arg: T7_Empty): Result<T7_Empty, T7_Empty> {
        return {tag: "Failure", error: new T7_Empty()};
    },
};

const throwingI1: I1 = {
    testCall(_arg: In): Result<Out, Err> {
        throw new Error("service error");
    },
    testCall2(_arg: T7_Empty): Result<T7_Empty, T7_Empty> {
        throw new Error("service error");
    },
};

const mockI2: I2 = {
    noErrCall(arg: In_No_Err): Out_No_Err {
        return new Out_No_Err(arg.value.toString());
    },
};

describe("I1 JSON wiring (Result)", () => {
    test("testCall returns Success", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall"};
        const inputJson = JSON.stringify(In.jsonCodec().encode(BaboonCodecContext.Default, new In()));
        const result = invokeJson_I1(method, inputJson, mockI1, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Success");
        if (result.tag === "Success") {
            const decoded = Out.jsonCodec().decode(BaboonCodecContext.Default, JSON.parse(result.value));
            expect(decoded.i00).toBe(42);
        }
    });

    test("testCall2 returns Success", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall2"};
        const inputJson = JSON.stringify(T7_Empty.jsonCodec().encode(BaboonCodecContext.Default, new T7_Empty()));
        const result = invokeJson_I1(method, inputJson, mockI1, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Success");
    });

    test("returns Failure for domain error", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall"};
        const inputJson = JSON.stringify(In.jsonCodec().encode(BaboonCodecContext.Default, new In()));
        const result = invokeJson_I1(method, inputJson, failingI1, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Failure");
    });

    test("returns Failure for unknown method", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "nonexistent"};
        const result = invokeJson_I1(method, "{}", mockI1, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Failure");
    });

    test("returns Failure for bad input", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall"};
        const result = invokeJson_I1(method, "not valid json!!", mockI1, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Failure");
    });

    test("returns Failure when service throws", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall"};
        const inputJson = JSON.stringify(In.jsonCodec().encode(BaboonCodecContext.Default, new In()));
        const result = invokeJson_I1(method, inputJson, throwingI1, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Failure");
    });
});

describe("I1 UEBA wiring (Result)", () => {
    test("testCall returns Success", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall"};
        const writer = new BaboonBinWriter();
        In.binCodec().encode(ctx, new In(), writer);
        const result = invokeUeba_I1(method, writer.toBytes(), mockI1, rt, ctx) as Result<Uint8Array, unknown>;
        expect(result.tag).toBe("Success");
        if (result.tag === "Success") {
            const reader = new BaboonBinReader(result.value);
            const decoded = Out.binCodec().decode(ctx, reader);
            expect(decoded.i00).toBe(42);
        }
    });

    test("returns Failure for unknown method", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "nonexistent"};
        const result = invokeUeba_I1(method, new Uint8Array(0), mockI1, rt, ctx) as Result<Uint8Array, unknown>;
        expect(result.tag).toBe("Failure");
    });

    test("returns Failure when service throws", () => {
        const method: BaboonMethodId = {serviceName: "I1", methodName: "testCall"};
        const writer = new BaboonBinWriter();
        In.binCodec().encode(ctx, new In(), writer);
        const result = invokeUeba_I1(method, writer.toBytes(), throwingI1, rt, ctx) as Result<Uint8Array, unknown>;
        expect(result.tag).toBe("Failure");
    });
});

describe("I2 JSON wiring (Result)", () => {
    test("noErrCall returns Success", () => {
        const method: BaboonMethodId = {serviceName: "I2", methodName: "noErrCall"};
        const inputJson = JSON.stringify(In_No_Err.jsonCodec().encode(BaboonCodecContext.Default, new In_No_Err(123)));
        const result = invokeJson_I2(method, inputJson, mockI2, rt, ctx) as Result<string, unknown>;
        expect(result.tag).toBe("Success");
        if (result.tag === "Success") {
            const decoded = Out_No_Err.jsonCodec().decode(BaboonCodecContext.Default, JSON.parse(result.value));
            expect(decoded.result).toBe("123");
        }
    });
});

describe("I2 UEBA wiring (Result)", () => {
    test("noErrCall returns Success", () => {
        const method: BaboonMethodId = {serviceName: "I2", methodName: "noErrCall"};
        const writer = new BaboonBinWriter();
        In_No_Err.binCodec().encode(ctx, new In_No_Err(456), writer);
        const result = invokeUeba_I2(method, writer.toBytes(), mockI2, rt, ctx) as Result<Uint8Array, unknown>;
        expect(result.tag).toBe("Success");
        if (result.tag === "Success") {
            const reader = new BaboonBinReader(result.value);
            const decoded = Out_No_Err.binCodec().decode(ctx, reader);
            expect(decoded.result).toBe("456");
        }
    });
});

File: test/dt-stub-wiring-overlay/test/wiring_test.dart
// Service-wiring + cross-domain muxer tests (no-errors service-result mode).
// Generated into this stub only by the `test-gen-dt-wiring` mdl action
// (rsync + codegen with --service-result-no-errors=true). Exercises the
// per-domain dispatchers (I1Wiring/I2Wiring) and the cross-domain
// JsonMuxer/UebaMuxer that routes by method.serviceId.
import 'dart:convert';
import 'dart:typed_data';

import 'package:baboon_runtime/baboon_runtime.dart';
import 'package:test/test.dart';

import '../lib/testpkg/pkg0/i1.dart';
import '../lib/testpkg/pkg0/i2.dart';
import '../lib/testpkg/pkg0/i1/testcall/in.dart' as i1in;
import '../lib/testpkg/pkg0/i1/testcall/out.dart' as i1out;
import '../lib/testpkg/pkg0/i2/noerrcall/in.dart' as i2in;
import '../lib/testpkg/pkg0/i2/noerrcall/out.dart' as i2out;
import '../lib/testpkg/pkg0/t7_empty.dart';

class MockI1 implements I1 {
  @override
  i1out.out testCall(i1in.in_ arg) => const i1out.out(i00: 42);
  @override
  T7_Empty testCall2(T7_Empty arg) => const T7_Empty();
}

class MockI2 implements I2 {
  @override
  i2out.out noErrCall(i2in.in_ arg) => i2out.out(result: 'result_${arg.value}');
}

void main() {
  final ctx = BaboonCodecContext.defaultCtx;

  String encodeI1In() => jsonEncode(i1in.in_JsonCodec.instance.encode(ctx, const i1in.in_()));
  String encodeI2In() => jsonEncode(i2in.in_JsonCodec.instance.encode(ctx, const i2in.in_(value: 123)));

  // ==================== Per-domain dispatch ====================

  test('I1Wiring.invokeJson routes testCall', () {
    final method = BaboonMethodId('I1', 'testCall');
    final result = I1Wiring.invokeJson(method, encodeI1In(), MockI1(), ctx);
    final decoded = i1out.out_JsonCodec.instance.decode(ctx, jsonDecode(result));
    expect(decoded.i00, 42);
  });

  test('I1Wiring.invokeJson throws NoMatchingMethod', () {
    final method = BaboonMethodId('I1', 'nonexistent');
    expect(
      () => I1Wiring.invokeJson(method, '{}', MockI1(), ctx),
      throwsA(predicate((e) => e is BaboonWiringException && e.error is NoMatchingMethod)),
    );
  });

  // ==================== Cross-domain Muxer ====================
  // A single muxer composes I1 and I2 and routes by method.serviceId.

  JsonMuxer<String> newJsonMuxer() =>
      JsonMuxer<String>([I1JsonService(MockI1()), I2JsonService(MockI2())]);
  UebaMuxer<Uint8List> newUebaMuxer() =>
      UebaMuxer<Uint8List>([I1UebaService(MockI1()), I2UebaService(MockI2())]);

  test('JsonMuxer routes to I1', () {
    final method = BaboonMethodId('I1', 'testCall');
    final result = newJsonMuxer().invoke(method, encodeI1In(), ctx);
    final decoded = i1out.out_JsonCodec.instance.decode(ctx, jsonDecode(result));
    expect(decoded.i00, 42);
  });

  test('JsonMuxer routes to I2', () {
    final method = BaboonMethodId('I2', 'noErrCall');
    final result = newJsonMuxer().invoke(method, encodeI2In(), ctx);
    final decoded = i2out.out_JsonCodec.instance.decode(ctx, jsonDecode(result));
    expect(decoded.result, 'result_123');
  });

  test('JsonMuxer throws NoMatchingService for unregistered service', () {
    final method = BaboonMethodId('Nonexistent', 'x');
    expect(
      () => newJsonMuxer().invoke(method, '{}', ctx),
      throwsA(predicate((e) => e is BaboonWiringException && e.error is NoMatchingService)),
    );
  });

  test('JsonMuxer throws DuplicateService on duplicate registration', () {
    expect(
      () => JsonMuxer<String>([I1JsonService(MockI1()), I1JsonService(MockI1())]),
      throwsA(predicate((e) => e is BaboonWiringException && e.error is DuplicateService)),
    );
  });

  test('UebaMuxer routes to I1', () {
    final method = BaboonMethodId('I1', 'testCall');
    final writer = BaboonBinTools.createWriter();
    i1in.in_UebaCodec.instance.encode(ctx, writer, const i1in.in_());
    final result = newUebaMuxer().invoke(method, writer.toBytes(), ctx);
    final decoded = i1out.out_UebaCodec.instance.decode(ctx, BaboonBinTools.createReader(result));
    expect(decoded.i00, 42);
  });

  test('UebaMuxer routes to I2', () {
    final method = BaboonMethodId('I2', 'noErrCall');
    final writer = BaboonBinTools.createWriter();
    i2in.in_UebaCodec.instance.encode(ctx, writer, const i2in.in_(value: 456));
    final result = newUebaMuxer().invoke(method, writer.toBytes(), ctx);
    final decoded = i2out.out_UebaCodec.instance.decode(ctx, BaboonBinTools.createReader(result));
    expect(decoded.result, 'result_456');
  });

  test('UebaMuxer throws NoMatchingService for unregistered service', () {
    final method = BaboonMethodId('Nonexistent', 'x');
    expect(
      () => newUebaMuxer().invoke(method, BaboonBinTools.createWriter().toBytes(), ctx),
      throwsA(predicate((e) => e is BaboonWiringException && e.error is NoMatchingService)),
    );
  });
}
